### PR TITLE
Update registration after every write

### LINKED ISSUE #2668: Connections between peers on a private testnet drop for no apparent reason
With Parity version `v1.4.0-unstable-271bcf4-20161009/x86_64-linux-gnu/rustc1.12.0` and a private testnet (spec file: https://gist.github.com/lgpawel/a6ba7660b778dc9775b00849abfc8be0) I experience unexpected disconnects. In both of the following examples, there are just two instances of Parity running parallel on the same machine, launched with `--bootnodes` pointing the instances to one another. For some reason (which may or may not be a separate issue), the numbers of connected peers are bigger than 1, as if there were multiple parallel connections between the instances.

In these logs the instances sit idle reporting two connections between them, then one of them drops, and after some time the other does too: https://gist.github.com/lgpawel/711033cfc0f4e0bd180d2e5397d3e91b

In these, the instances sit idle, apparently stable with three connections between them, then I launch mining for both of them (explicitly limiting it to respectively 1 and 2 cores of a four-core processor) and immediately the connections are dropped, this time with much more verbose output: https://gist.github.com/lgpawel/c7c093325a533388d961e9e294849290

### PR DESCRIPTION
**Pull request by arkpar**
Should fix #2668

### CODE REVIEW

File: util/network/src/connection.rs [hunk 1]
```diff
@@ -158,8 +158,8 @@ impl<Socket: GenericSocket> GenericConnection<Socket> {
 			}
 			if self.send_queue.is_empty() {
 				self.interest.remove(Ready::writable());
-				io.update_registration(self.token)?;
 			}
+			io.update_registration(self.token)?;
```
Comment (arkpar) [inline util/network/src/connection.rs]: This fixes a race condition when the sending queue is being processed and another packet is added to the queue. The registration might have not been updated in such case.

Comment (rphmeier) [inline util/network/src/connection.rs]: Not too clear on the mio specifics, but does it cause any issues to `reregister` interest multiple times before a `timeout` or `notify` event is handled? I thought that was the original purpose of waiting until the send queue was empty.

Comment (rphmeier) [inline util/network/src/connection.rs]: Also, how could there be such a race condition? Uou have mutable access to `self` here, nothing else can access it.

Comment (arkpar) [inline util/network/src/connection.rs]: Calling `reregister` multiple times is perfectly fine.

**Pull request merged**